src/Symfony/Component/Lock/Store/PdoStore.php

```diff
@@ -0,0 +1,347 @@
+        } elseif ($connOrDsn instanceof Connection) {
+            $this->conn = $connOrDsn;
+        } elseif (is_string($connOrDsn)) {
+            $this->dsn = $connOrDsn;
+        } else {
+            throw new InvalidArgumentException(sprintf('"%s" requires PDO or Doctrine\DBAL\Connection instance or DSN string as first argument, "%s" given.', __CLASS__, is_object($connOrDsn) ? get_class($connOrDsn) : gettype($connOrDsn)));
+        }
+
+        $this->table = $options['db_table'] ?? $this->table;
+        $this->idCol = $options['db_id_col'] ?? $this->idCol;
+        $this->tokenCol = $options['db_token_col'] ?? $this->tokenCol;
+        $this->expirationCol = $options['db_expiration_col'] ?? $this->expirationCol;
+        $this->username = $options['db_username'] ?? $this->username;
+        $this->password = $options['db_password'] ?? $this->password;
+        $this->connectionOptions = $options['db_connection_options'] ?? $this->connectionOptions;
+
+        $this->drift = $drift;
+        $this->gcProbability= $gcProbability;
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key)
+    {
+        $key->reduceLifetime($this->initialTtl);
+
+        $sql = "INSERT INTO $this->table ($this->idCol, $this->tokenCol, $this->expirationCol) VALUES (:id, :token, :expiration)";
+        $stmt = $this->getConnection()->prepare($sql);
+
+        $stmt->bindValue(':id', $this->getHashedKey($key));
+        $stmt->bindValue(':token', $this->getToken($key));
+        $stmt->bindValue(':expiration', time() + $this->initialTtl + $this->drift, \PDO::PARAM_INT);
+
+        try {
+            $stmt->execute();
+            if ($key->isExpired()) {
+                throw new LockExpiredException(sprintf('Failed to put off the expiration of the "%s" lock within the specified time.', $key));
+            }
+
+            return;
+        } catch (UniqueConstraintViolationException $e) {
+            // the lock is already acquired. It could be us. Let's try to put off.
+            $this->putOffExpiration($key, $this->initialTtl);
+        } catch (\PDOException $e) {
+            // the lock is already acquired. It could be us. Let's try to put off.
+            $this->putOffExpiration($key, $this->initialTtl);
+        }
+
+        if ($key->isExpired()) {
+            throw new LockExpiredException(sprintf('Failed to store the "%s" lock.', $key));
+        }
+
+        if ($this->gcProbability > 0 && ($this->gcProbability === 1.0 || (random_int(0, PHP_INT_MAX) / PHP_INT_MAX) <= $this->gcProbability)) {
+            $this->prune();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function waitAndSave(Key $key)
+    {
+        throw new NotSupportedException(sprintf('The store "%s" does not supports blocking locks.', __CLASS__));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function putOffExpiration(Key $key, $ttl)
+    {
+        if ($ttl < 1) {
```

Comment:
I wonder if we should handle TTL < 1 sec ?
If yes, it requires to manage micro timestamp in all drivers which is harder than a simple integer.

But does it worth it, it would mean that App server and Database server are time synchronized with 1 second precision.

Reply:
To avoid time synchronisation issues, we can rely on the database time only.

```sql
UPDATE $this->table
SET
  $this->expirationCol = UTC_TIMESTAMP() + $ttl,
  $this->tokenCol = :token
WHERE $this->idCol = :id
AND ($this->tokenCol = :token OR $this->expirationCol <= UTC_TIMESTAMP())
```

Reply:
:+1: would be far better

Reply:
@jderusse waiting for you here, isn't it?